Collision monitor (#2982)
ros-planning/navigation2: nav2_collision_monitor/README.md

# Nav2 Collision Monitor

> Collision Monitor - is an independent layer in Nav2 providing an additional level of robot safety assurance.

Reviewer: The Collision Monitor is a node providing an additional level of robot safety. It performs several collision avoidance related tasks using incoming data from the sensors, bypassing the costmap and trajectory planners, to monitor for and prevent potential collisions at the emergency-stop level.

This is analogous to safety sensor and hardware features; take in laser scans from a real-time certified safety scanner, detect if there is to be an imminent collision in a configurable bounding box, and either emergency-stop the certified robot controller or slow the robot to avoid such collision. However, this node is done at the CPU level with any form of sensor. As such, this does not provide hard real-time safety certifications, but uses the same types of techniques with the same types of data for users that do not have safety-rated laser sensors, safety-rated controllers, or wish to use any type of data input (e.g. pointclouds from depth or stereo or range sensors).  

This is a useful and integral part of large heavy industrial robots, or robots moving with high velocities, around people or other dynamic agents (e.g. other robots) as a safety mechanism for high-response emergency stopping. The costmaps / trajectory planners will handle most situations, but this is to handle obstacles that virtually appear out of no where (from the robot's perspective) or approach the robot at such high speed it needs to immediately stop to prevent collision.  

---

Do not say things like "assurance" <-- we're not assuring anything

Author: Great introduction update. Thanks! Update the latest `README.md` with it.

It allows performing robot collision avoidance by monitoring the obstacles in surrounding environments laying in collision proximity to the robot.

This is extremely useful and integral part of large heavy industrial robots operating in the facilities, or of robots moving with high velocities, where the collision with any obstacle might cause unpredictable severe consequences.

> 

Reviewer: I might suggest:
- Putting the robot into another map or into a less cluttered part of the map (e.g. off to the right) so when you take the screen shot, its not already in collision with something
- Use that zoomed in image from the right side 
- Add labels to the 3 polygons for their uses
- Remove the particle filter arrows 
- Making the slow down polygon all around the robot like you had as the square in the tutorial. I liked that idea. The stop zone should be just in front of the robot for emergencies, but generally slowing in proximity to obstacles is nice.

> 

Author: Updated the picture according to your suggestion

## Features

> Collision Monitor operates by means of areas exposed around the robot and moving with it.

Reviewer: The Collision Monitor uses polygons relative the robot's base frame origin to define "zones". Data that fall into these zones trigger an operation depending on the model being used. A given instance of the Collision Monitor can have many zones with different models at the same time. When multiple zones trigger at once, the most aggressive one is used (e.g. stop > slow 50% > slow 10%).

Author: Done in latest changes.

The obstacles fell into such area are forcing robot to stop, operate slowly or even fluently approach an obstacle with a constant time to the collision.

The following models of safety behaviors are employed by Collision Monitor:

> * **Stop model**: Define a safety area surrounding the robot and a point threshold. If more that `N` obstacle points appear inside this area, stop the robot until the obstacles will disappear.

Reviewer: ```suggestion
* **Stop model**: Define a zone and a point threshold. If more that `N` obstacle points appear inside this area, stop the robot until the obstacles will disappear.
```

> * **Slowdown model**: Define a safety area around the robot and slow the maximum speed for a `%S` percent, if more than `N` points will appear inside the area.

Reviewer: ```suggestion
* **Slowdown model**: Define a zone around the robot and slow the maximum speed for a `%S` percent, if more than `N` points will appear inside the area.
```

> * **Approach model**: With the current robot speed, estimate the time to collision to points obtained from sensors. If the time is less than `M` seconds (0.5, 2, 5, etc...), slow the robot until it will be equal to `M` seconds. The effect here would be to keep the robot always `M` seconds from a collision and continuously scale down its speeds.

Reviewer: ```suggestion
* **Approach model**: Using the current robot speed, estimate the time to collision to sensor data. If the time is less than `M` seconds (0.5, 2, 5, etc...), the robot will slow such that it is now at least `M` seconds to collision. The effect here would be to keep the robot always `M` seconds from any collision.
```

The safety area around the robot can take the following shapes:

> * Arbitrary user-defined polygon around the robot for usage in stop and slowdown models.

Reviewer: ```suggestion
* Arbitrary user-defined polygon relative to the robot base frame.
```

> * Robot footprint polygon, which is used in the approach behavior model only.

Reviewer: ```suggestion
* Robot footprint polygon, which is used in the approach behavior model only. Will use the footprint topic to allow it to be dynamically adjusted over time.
```

> * Circle, used in all models: for stop and slowdown models as a safety area, for approach model as a robot footprint. Circle is made for the best performance and could be used in the cases where the safety area or robot could be approximated by round shape.

Reviewer: ```suggestion
* Circle: is made for the best performance and could be used in the cases where the safety area or robot could be approximated by round shape.
```

Move circle before the footprint so the general ones are together first.

Author: Changed and moved up in https://github.com/ros-planning/navigation2/pull/2982/files#diff-95f4c0b773dca7c45bd03829c788a4d6ed3ddf65394702457138ffc35a3355c2R29.

> NOTE: Although safety behavior models are not intended to be used simultaneously (e.g. stop model should not be crossed with approach one), it is not prohibited to. Collision Monitor allows setting simultaneously multiple shapes with different behavior models. This typically could be useful to have a small stop area and larger slowdown bounding box to warn the robot from a collision without termination of operation.

Author: [Done](https://github.com/ros-planning/navigation2/pull/2982/commits/e58413d79bc55b383be13969f0ef654d6bef7237#diff-95f4c0b773dca7c45bd03829c788a4d6ed3ddf65394702457138ffc35a3355c2L25)

The obstacle points are being obtained from different data sources. Collision Monitor is subscribed to:

* Laser scanners (`sensor_msgs::msg::LaserScan` messages)
* PointClouds (`sensor_msgs::msg::PointCloud2` messages)
* IR/Sonars (`sensor_msgs::msg::Range` messages, not implemented yet)

## Design

> Since Collision Monitor is designed to operate as an independent safety node, Nav2 stack has no knowledge about it.

Reviewer: ```suggestion
The Collision Monitor is designed to operate below Nav2 as an independent safety node.
```

Remove the next line starting with "It is laying..." that is redundant

It is laying under the stack and operating after all necessary decisions were made by Nav2.
This is achieved through remapped `cmd_vel` topic, going out from a Controller.

> The following diagram is showing the high-level design of Collision Monitor module. All shapes (Polygons and Circles) are derived from base `Polygon` class, so without loss of generality we can call them as polygons. Subscribed footprint is also having the same properties as other polygons, but it is being obtained from `nav2_costmap_2d::FootprintSubscriber`.

Reviewer: ```suggestion
The following diagram is showing the high-level design of Collision Monitor module. All shapes (Polygons and Circles) are derived from base `Polygon` class, so without loss of generality we can call them as polygons. Subscribed footprint is also having the same properties as other polygons, but it is being obtained a footprint topic for the Approach Model. 
```

![HDL.png](doc/HLD.png)

## Configuration

Detailed configuration parameters, their description and how to setup a Collision Monitor could be found at its [Configuration Guide](https://navigation.ros.org/configuration/packages/configuring-collision-monitor.html) and [Using Collision Monitor tutorial](https://navigation.ros.org/tutorials/docs/using_collision_monitor.html) pages.

## Metrics

Designed to be used in wide variety of robots (incl. moving fast) and have a high level of reliability, Collision Monitor node should operate at fast rates.
Typical one frame processing time is ~4-5ms for laser scanner (with 360 points) and ~4-80ms for PointClouds (having 24K points).
The table below represents the details of operating times for different behavior models and shapes:

| | Stop/Slowdown model, Polygon area | Stop/Slowdown model, Circle area | Approach model, Polygon footprint | Approach model, Circle footprint |
|-|-----------------------------------|----------------------------------|-----------------------------------|----------------------------------|
| LaserScan (360 points) processing time, ms  | 4.09 | 4.08 | 4.98  | 4.29  |

> | PointCloud (24K points) processing time, ms | 4.13 | 3.76 | 77.92 | 11.43 |

Reviewer: Is there a way to appreciably speed up the approach model polygon? Maybe a different algorithm or algorithm optimizations?

Reviewer: Try to embed the `getPointInside` inside of `getPointsInside`?

Reviewer: Multithreading?

Author: Yes, I thought about ways of optimization.

[Here](https://erich.realtimerendering.com/ptinpoly/) is the work of comparing the performance for different algorithms about point-in-polygon solution. Among basic algorithms, raycrossing algorithm is the fastest one. There is an optimization, called as "MacMartin" used for large polygons, which speeds up the raycrossing algorithms on x1.8 times.
The best way of enhance algorithms: is a grid algorithm, quantizing the map and making "inside", "non-inside" and "indeterminate" cells. According to the work, this algorithm gives best performance, and thus more promising as an optimization point.

> Try to embed the getPointInside inside of getPointsInside?

This function was already inlined locally: `inline bool Polygon::isPointInside(const Point & point) const`, and confirmed by binary disassembling.

The multithreading is another way of optimization. Currently, the motion is being simulated by small steps. But it could being calculated the exact position of robot pose after `t` time (since we have `theta` and linear velocity, this making arcs; and and then we need to make a transform from robot's velocity frame down to base frame). Knowing the exact robot position in any time frame, we could parallel the whole calculation loop e.g. by giving odd iterations to one thread, and even ones - to another. All iterations will be independent from each other. This is another point of optimization could be made there.

Author: Update: the multithreading could be parallelized by points, instead of simulation steps. This makes it easiest to implement giving tangible benefits on multi-CPU systems.

Reviewer: What would a default of `0.1` for simulation timestep yield? I think 0.1s timestep increments are plenty small and are 5x larger than currently set.

Author: That is definitely makes sense. The Footprint Approach time was decreased from ~78 -> ~20 ms. Re-measured the peroformance and [updated](https://github.com/ros-planning/navigation2/pull/2982/commits/a4639aac3380a3bdbf918895319cb0e2721f680f) the README accordingly.

The following notes could be made:

>  * Due to faster algorithms, circle shapes are preferred for the approach behavior models.

Reviewer: ```suggestion
 * Due to sheer speed, circle shapes are preferred for the approach behavior models if you can approximately model your robot as circular.
```

>  * More points mean lower performance. Pointclouds could be sparsed before Collision Monitor node for the systems with lack of computational resources.

Reviewer: ```suggestion
 * More points mean lower performance. Pointclouds could be culled or filtered before the Collision Monitor to improve performance. 
```
